pkg/controller/podgc/gc_controller_test.go

```diff
@@ -657,6 +662,181 @@ func TestGCTerminating(t *testing.T) {
+			},
+		},
+	}
+
+	testCases := []struct {
+		name                          string
+		node                          *v1.Node
+		pod                           *v1.Pod
+		expectedPatchedPod            *v1.Pod
+		expectedDeleteAction          *clienttesting.DeleteActionImpl
+		enablePodDisruptionConditions bool
+	}{
+		{
+			name: "orphaned pod should be deleted",
+			node: testNode,
+			pod: &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: "default",
+					Name:      "testPod",
+				},
+				Spec: v1.PodSpec{
+					NodeName: "deletedNode",
+				},
+				Status: v1.PodStatus{
+					Phase: v1.PodRunning,
+					Conditions: []v1.PodCondition{
+						{
+							Type:   v1.PodReady,
+							Status: v1.ConditionTrue,
+						},
+					},
+				},
+			},
+			expectedPatchedPod: &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: "default",
+					Name:      "testPod",
+				},
+				Spec: v1.PodSpec{
+					NodeName: "deletedNode",
+				},
+				Status: v1.PodStatus{
+					Phase: v1.PodFailed,
+					Conditions: []v1.PodCondition{
+						{
+							Type:    v1.DisruptionTarget,
+							Status:  v1.ConditionTrue,
+							Reason:  "DeletionByPodGC",
+							Message: "PodGC: node no longer exists",
+						},
+						{
+							Type:   v1.PodReady,
+							Status: v1.ConditionTrue,
+						},
+					},
+				},
+			},
+			expectedDeleteAction: &clienttesting.DeleteActionImpl{
+				Name:          "testPod",
+				DeleteOptions: metav1.DeleteOptions{GracePeriodSeconds: pointer.Int64(0)},
+			},
+			enablePodDisruptionConditions: true,
+		},
+		{
+			name: "not orphaned pod should not be deleted",
+			node: testNode,
+			pod: &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: "default",
+					Name:      "testPod",
+				},
+				Spec: v1.PodSpec{
+					NodeName: "testNode",
+				},
+				Status: v1.PodStatus{
+					Phase: v1.PodRunning,
+					Conditions: []v1.PodCondition{
+						{
+							Type:   v1.PodReady,
+							Status: v1.ConditionTrue,
+						},
+					},
+				},
+			},
+			expectedPatchedPod: &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: "default",
+					Name:      "testPod",
+				},
+				Spec: v1.PodSpec{
+					NodeName: "testNode",
+				},
+				Status: v1.PodStatus{
+					Phase: v1.PodRunning,
+					Conditions: []v1.PodCondition{
+						{
+							Type:   v1.PodReady,
+							Status: v1.ConditionTrue,
+						},
+					},
+				},
+			},
+			expectedDeleteAction:          nil,
+			enablePodDisruptionConditions: true,
+		},
+	}
+
+	for _, test := range testCases {
+		t.Run(test.name, func(t *testing.T) {
+			defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.PodDisruptionConditions, test.enablePodDisruptionConditions)()
+
+			nodes := []*v1.Node{test.node}
+			pods := []*v1.Pod{test.pod}
+
+			client := setupNewSimpleClient(nodes, pods)
+			gcc, _, _ := NewFromClient(client, -1)
+			gcc.quarantineTime = time.Duration(-1)
+			gcc.gcOrphaned(context.TODO(), pods, nodes)
```

Comment:
I'm wondering if it makes more sense to call the generic `gcc.gc()` such as here: https://github.com/kubernetes/kubernetes/blob/7e5506de8d44d173ff18c4bfc3a9e62e5ada3161/pkg/controller/podgc/gc_controller_test.go#L372. Then, we could reuse this test case for other gc scenarios, other than just orphan pods.

Reply:
Changed

Reply:
I am not sure how to use `gcc.gc`, since you won't be able to pass in the `pods` and `nodes`. Can you provide some guidance into how to do that?

Reply:
A method like here should work: https://github.com/kubernetes/kubernetes/blob/7e5506de8d44d173ff18c4bfc3a9e62e5ada3161/pkg/controller/podgc/gc_controller_test.go#L349-L369 using informers at the client level. 

Yes, it is more glue code, but it is more generic and would allow use to test other scenarios than orphan pods in the same test case, dedicated to testing the patch.

Reply:
Thanks! Learned something new :)

Reply:
Updated